include/llama/Array.hpp

```diff
@@ -89,6 +89,56 @@ namespace llama
 
     template <typename First, typename... Args>
     Array(First, Args... args) -> Array<First, sizeof...(Args) + 1>;
+
+    template <typename T, std::size_t N>
+    LLAMA_FN_HOST_ACC_INLINE constexpr auto push_front(Array<T, N> a, T v) -> Array<T, N + 1>
+    {
+        Array<T, N + 1> r{};
+        r[0] = v;
+        for (auto i = 0; i < N; i++)
+            r[i + 1] = a[i];
+        return r;
+    }
+
+    template <typename T, std::size_t N>
+    LLAMA_FN_HOST_ACC_INLINE constexpr auto push_back(Array<T, N> a, T v) -> Array<T, N + 1>
+    {
+        Array<T, N + 1> r{};
+        for (auto i = 0; i < N; i++)
+            r[i] = a[i];
+        r[N] = v;
+        return r;
+    }
+
+    template <typename T, std::size_t N>
+    LLAMA_FN_HOST_ACC_INLINE constexpr auto pop_back(Array<T, N> a) -> Array<T, N - 1>
+    {
+        Array<T, N - 1> r{};
+        for (auto i = 0; i < N - 1; i++)
+            r[i] = a[i];
+        return r;
+    }
+
+    template <typename T>
+    LLAMA_FN_HOST_ACC_INLINE constexpr auto pop_back(Array<T, 0> a) -> Array<T, 0>
```

Comment:
What are the intended semantics here? To me this sounds like a case for `static_assert`.

Reply:
I pulled this out of a different PR, where I need this precise semantic. But we could argue it should be a differently named function. I will probably change the name then.

Reply:
I found a way to not need this special case. I removed it.